rust/blockstore/src/arrow/blockfile.rs

```diff
@@ -478,42 +482,41 @@ impl<'me, K: ArrowReadableKey<'me> + Into<KeyWrapper>, V: ArrowReadableValue<'me
     }
 
     // Returns all Arrow records in the specified range.
-    pub(crate) async fn get_range<'prefix, PrefixRange, KeyRange>(
+    pub(crate) fn get_range<'prefix, PrefixRange, KeyRange>(
         &'me self,
         prefix_range: PrefixRange,
         key_range: KeyRange,
-    ) -> Result<Vec<(K, V)>, Box<dyn ChromaError>>
+    ) -> impl Stream<Item = Result<(K, V), Box<dyn ChromaError>>> + Send + 'me
     where
-        PrefixRange: RangeBounds<&'prefix str> + Clone,
-        KeyRange: RangeBounds<K> + Clone,
+        PrefixRange: RangeBounds<&'prefix str> + Clone + Send + 'me,
+        KeyRange: RangeBounds<K> + Clone + Send + 'me,
+        K: Sync,
+        V: Sync,
     {
-        let block_ids = self
-            .root
-            .sparse_index
-            .get_block_ids_range(prefix_range.clone(), key_range.clone());
-
-        let mut result: Vec<(K, V)> = vec![];
-        for block_id in block_ids {
-            let block_opt = match self.get_block(block_id).await {
-                Ok(Some(block)) => Some(block),
-                Ok(None) => {
-                    return Err(Box::new(ArrowBlockfileError::BlockNotFound));
-                }
-                Err(e) => {
-                    return Err(Box::new(e));
-                }
-            };
-
-            let block = match block_opt {
-                Some(b) => b,
-                None => {
-                    return Err(Box::new(ArrowBlockfileError::BlockNotFound));
-                }
-            };
-            result.extend(block.get_range(prefix_range.clone(), key_range.clone()));
-        }
-
-        Ok(result)
+        futures::stream::iter(
+            self.root
+                .sparse_index
+                .get_block_ids_range(prefix_range.clone(), key_range.clone())
+                .into_iter()
+                .map(Ok),
+        )
+        .try_filter_map(move |block_id| async move {
+            match self.get_block(block_id).await {
+                Ok(Some(block)) => Ok(Some(block)),
+                Ok(None) => Err(Box::new(ArrowBlockfileError::BlockNotFound)),
+                Err(e) => Err(Box::new(ArrowBlockfileError::BlockFetchError(e))),
+            }
+        })
+        .map(move |block| match block {
+            Ok(block) => futures::stream::iter(
+                block
+                    .get_range::<K, V, _, _>(prefix_range.clone(), key_range.clone())
+                    .map(Ok),
+            )
+            .boxed(),
+            Err(e) => futures::stream::once(async { Err(e as Box<dyn ChromaError>) }).boxed(),
```

Comment:
I think this boxing is unavoidable, couldn't figure out a clever workaround (maybe creating a custom Stream type?)